docs/source/design/python_imports.rst

```diff
@@ -0,0 +1,61 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_design_python_imports:
+
+************************
+Python import statements
+************************
+
+.. topic:: Specification scope and status
+
+   This specification describes the current (albeit incomplete) implementation.
+
+The following rules apply to any ``import`` statement in the code base:
+
+- All imports *must* be absolute.
```

Comment:
Draft: yet to finish.

As many (if not all) categorical decisions like this, they tend to need exceptions to the rule to make them pragmatic and efficient.

Formatting aside, here is a suggestion:

```suggestion
- All imports *must* be absolute unless modules are *semantically bound* in the hierarchy.

Instances of *semantically bound* modules:

- Tests: `test_<submodule>.py` should use relative form `from ..submodule import ...` to import functionality subject to testing.

Rationale: to facilitate selective testing etc, tests for functionality in the `submodule` reside under `submodule.tests`. Use of relative imports here binds together location of the functionality and the tests allowing for 1). a corresponding module and its tests to be moved within codebase without requiring any code change; 2). easy and unambiguous identification of what functionality is tested without doing visual names matching.

Note: even though `datalad.tests.utils` module resides under `tests`, absolute import should be used to import it for unless in the `datalad.tests.test_tests_utils` which tests functionality of that module.

- Import of base interfaces or specific implementations within some interface: ... 
```

Reply:
re tests: good point, thx! I did not think much of the tests (in often keep them unchanged in my edits), but it makes sense to handle them here. I am curious what other scenarios of "semantically bound" you will list.

Reply:
extended (yet to be contracted, added examples to support) above... 

<details>
<summary>here is an example how it would look e.g. in datalad/downloaders/providers.py</summary> 

```python
import os
import re
from collections import OrderedDict
from glob import glob
from logging import getLogger
from os.path import dirname, abspath, join as pathjoin
from urllib.parse import urlparse

from datalad.interface.common_cfg import dirs
from datalad.support.configparserinc import SafeConfigParserWithIncludes
from datalad.support.external_versions import external_versions
from datalad.support.network import RI
from datalad.support import path
from datalad.utils import auto_repr
from datalad.utils import ensure_list_from_str
from datalad.utils import get_dataset_root

from .base import NoneAuthenticator, NotImplementedAuthenticator
from .credentials import CREDENTIAL_TYPES
from .http import (
    HTMLFormAuthenticator,
    HTTPAnonBearerTokenAuthenticator,
    HTTPAuthAuthenticator,
    HTTPBasicAuthAuthenticator,
    HTTPDigestAuthAuthenticator,
    HTTPBearerTokenAuthenticator,
    HTTPDownloader,
)
from .s3 import (
    S3Authenticator,
    S3Downloader,
)
from .shub import SHubDownloader

```
</details>

Reply:
Just FTR, an obnoxious unification helper for a part of it
```
re='from \.\.\.*\(cmdline\|dochelpers\|downloaders\|support\|ui\)'; git grep -l "$re" | grep -v 'datalad/__init__.py' | xargs sed -i -e "s,$re, from datalad.\1,g"
```

note: any unification we do - we better do it shortly before 0.15.0 release.

Reply:
<details>
<summary>output of `python -m isort -m3 -fgw 2 -tc datalad/downloaders/providers.py; git diff` on that example I gave above makes it perfect:</summary> 

```diff
diff --git a/datalad/downloaders/providers.py b/datalad/downloaders/providers.py
index b0554cb45..d42b547d8 100644
--- a/datalad/downloaders/providers.py
+++ b/datalad/downloaders/providers.py
@@ -14,27 +14,36 @@ import re
 from collections import OrderedDict
 from glob import glob
 from logging import getLogger
-from os.path import dirname, abspath, join as pathjoin
+from os.path import (
+    abspath,
+    dirname,
+)
+from os.path import join as pathjoin
 from urllib.parse import urlparse
 
 from datalad.interface.common_cfg import dirs
+from datalad.support import path
 from datalad.support.configparserinc import SafeConfigParserWithIncludes
 from datalad.support.external_versions import external_versions
 from datalad.support.network import RI
-from datalad.support import path
-from datalad.utils import auto_repr
-from datalad.utils import ensure_list_from_str
-from datalad.utils import get_dataset_root
+from datalad.utils import (
+    auto_repr,
+    ensure_list_from_str,
+    get_dataset_root,
+)
 
-from .base import NoneAuthenticator, NotImplementedAuthenticator
+from .base import (
+    NoneAuthenticator,
+    NotImplementedAuthenticator,
+)
 from .credentials import CREDENTIAL_TYPES
 from .http import (
     HTMLFormAuthenticator,
     HTTPAnonBearerTokenAuthenticator,
     HTTPAuthAuthenticator,
     HTTPBasicAuthAuthenticator,
-    HTTPDigestAuthAuthenticator,
     HTTPBearerTokenAuthenticator,
+    HTTPDigestAuthAuthenticator,
     HTTPDownloader,
 )
 from .s3 import (
@@ -43,7 +52,6 @@ from .s3 import (
 )
 from .shub import SHubDownloader
 
-
 lgr = getLogger('datalad.downloaders.providers')
 
 # dict to bind authentication_type's to authenticator classes

```
</details>

Reply:
OK, I can easily get behind the test-scenario. If code is moved, tests will move too. I have done that dozens of times.

I do not get behind the other two scenarios. In particular the annex/gitrepo examples. When I created the recent Repo base classes, I had to RF exactly this kind of "bound" relationship. They are still "bound", by means of subclasses, but not as a monolithic unit.

I think the way I look at it is this:

> Relative imports are desirable/OK, if and only if code is split into multiple files for some technical reason, but semantically belongs to the same place (e.g., code and its tests), i.e. they could life in the same file.

For Git v AnnexRepo it is *one* possiblity to consider them a bound tandem of base classes, but it is just as meaningful to strictly separate annex-related code into a different part of an alternative source tree.

That being said, if we can agree on a short statement, I think that it could be interpreted in a way that also covers the other two scenarios you put forth. I would leave it to the respective developer what is considered a single semantic unit, and where to use relative imports. And I would grant the same freedom to the developer that breaks this unit when moving code and converts the respective imports to absolute ones ;-)

So what about this:

```suggestion
- All imports *must* be absolute, unless they import individual pieces of an integrated code component that is only split across several source code files for technical or organizational reasons.
```

I think this conditions covers all three scenarios you listed.

It is useful to remind ourselves at this point that the aim of these design documents is to provide guidance and practical advise to make it easy to do the "right thing", not to develop enforceable laws ;-)

Reply:
1. 
I agree, that "the other scenarios" are at the very least less convincing, than the test scenario. Any refactoring can and often will break those relationships of locations, making the relative imports more, not less troublesome in my view, since you'd need to figure out which ones would stay that way and which ones need to be changed to absolute ones in order to stay consistent with those principles. Furthermore, for me it's the opposite of "obfuscating code readability": I find it harder to read. Needs constant awareness where I am and what else is in that module. To me that is in fact only relevant when refactoring and that's the point where - in those scenarios - the relative imports are a pain not a help in my view.

2.
I don't even really buy the test scenario, although I'm willing to accept it, if insisted upon. I think the entire "move the code easily" argument is flawed (in any scenario). It may be, that I can easily move `submodule/*.py` and `submodule/tests/*.py` together w/o changing imports in the latter. However, if I do that, I'll have to adjust the imports for `submodule` through the code base anyway (otherwise it's unused and should be deleted rather than moved). Whether or not it's one more file doesn't matter at all. Furthermore import adjustments for movement of a module are easiest when everything absolute, since it's one simple replacement everywhere - within and outside of the module. And finally: I don't know the realm of vi plugins, but for everyone using something like PyCharm, it's "right click, left click, type in target, enter" - if everything is absolute. If not, then I need to go through the changes done by this and find which ones of the adjustments I need to turn into relative imports again. So, to me even the relative imports in tests are causing more effort to move a module, not less.

In conclusion: I'd be more happy with absolute everywhere - no exception. However, if there remains insistence on exceptions, I'm fine with the new wording suggested and not willing to fight over it really. It's not important to me, just an annoyance. :)

Reply:
> Whether or not it's one more file doesn't matter at all.

a minor point for defense: depending on the size/content of the file, and how move/adjustment is done (in a single commit or two commits) it would matter for `git` to identify the `mv` (AFAIK commit itself does not record that something is moved -- it is deduced "after-fact", I could be wrong), which might be crucial for seamless merging of `maint` into `master` whenever file was moved in `master. Introducing changes to the moved file in the same commit for imports (as any fancy IDE would do) should generally be "ok" since `git` allows for some "fuzz" in deducing the `mv`. Although `git` is smart that way, making it more obvious (no changes to moved file whatsoever) would assist `git` in figuring out the `mv` or might even become instrumental, if the file  is just imports and thus most of the content would be changed.

side note, unrelated to imports: for such `mv` commits, we **must not** introduce a shim in the old location in the same commit. That would make it even harder for `git` to deduce the move.  Unfortunate truth, is that ATM with `git 2.32.0` it still "does not matter" in that even with 2 commits, `git` would fail to identify that change introduced in original branch in original location, should be applied to the moved location, and not the shim.  But I am hopeful that `git` would get smarter ;)

<details>
<summary>FWIW -- demo reproducer. As long as shim is there -- `git merge` would fail :-/ </summary> 

```bash
#!/bin/bash

export PS4='> '
set -eu
cd "$(mktemp -d ${TMPDIR:-/tmp}/ann-XXXXXXX)"

set -x

git init 

mkdir sub1

cat >| sub1/mod1.py << EOF

from datalad.bugiwoogie import smth

from datalad.base import blah
from datalad.mod1.here import bleh


def test_blah():
    assert blah
EOF

git add sub1/mod1.py
git commit -m initialaddsub1

git checkout -b maint
sed -i -e 's,assert,better_assert,g' sub1/mod1.py
git commit -m 'fixed a bug in maint' -a

git checkout -
mkdir sub2
git mv sub1/* sub2/

# add change or not
# sed -i -e 's,sub1,sub2,g' sub2/*

git commit -m 'moved and adjusted sub1 content uder sub2 in master'

cat >| sub1/mod1.py << EOF
# we moved into sub1
EOF
git add sub1/mod1.py

git commit -m 'added placeholder under sub1 in master'

git merge -m 'merge maint after move' maint

```
</details>